### ISSUE TITLE
Template bug - relying on a parameter to have a `/` in front of it

### ISSUE THREAD

**Issue by Temikus** [buildkite/elastic-ci-stack-for-aws#812]
So AWS SSM parameter doesn't necessarily has to have a leading slash to it, e.g. `buildkite-token` is a perfectly valid SSM Parameter.

However, the AWS stack template seems to assume otherwise:
https://github.com/buildkite/elastic-ci-stack-for-aws/blob/master/templates/aws-stack.yml#L707

Leading to invalid references in roles, like:
```
"Resource": "arn:aws:ssm:ap-southeast-2:103639821168:parameterbuildkite-token-default"
                                                                                                                   ^^^
``` 
, instead of:
```
"Resource": "arn:aws:ssm:ap-southeast-2:103639821168:parameter/buildkite-token-default"
                                                                                                                   ^^^
```

**Comment by Temikus** [comment]
Here's a repro gist with my terraform template: https://gist.github.com/Temikus/6c485710fbc33ccbefe7a4a444605d65

**Comment by jufemaiz** [comment]
As per [Slack](https://buildkite-community.slack.com/archives/C19BG77H9/p1615799799014600):

I think the outcome here is going to be better documentation due to ssm requirements around parameter store.
Anything with a path must have a leading slash. With no path, that requirements is not present

> Fully qualified: When you create or reference a parameter in a hierarchy, you must include a leading forward slash character (`/`) . When you reference a parameter that is part of a hierarchy, you must specify the entire hierarchy path including the initial slash (`/`).
> 
> * Fully qualified parameter names: `MyParameter1`, `/MyParameter2`, `/Dev/Production/East/Project-ABC/MyParameter`
> * Not fully qualified parameter name: `MyParameter3/L1`

https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-paramstore-su-create.html

So a key challenge here is if Cloudformation can identify if there is a '/' in the provided path or not. If there is, don't join with a slash, if there is not join with a slash.

The alternative would be ensuring the documentation emphasises the expectation of a leading slash irrespective of wether the `name` attribute needs it or not (that is, whether it is part of a hierarchy).

**Comment by yob** [comment]
Sounds like you're more familiar with SSM ParameterStore than I am @jufemaiz.

We could change the `BuildkiteAgentTokenParameterStorePath` stack parameter to validate that the value always starts with a `/` - at least users would get a decent error message. Is `buildkite-token-default` equal to `/buildkite-token-default` for @Temikus though?

If not, and we need to append the slash only when constructing the ARN. Hmmm. Maybe there's a hard-to-read combination of `split()` and `join()` that can conditionally add the `/`? :thinking:

**Comment by xiaket** [comment]
I think it is better to document this behaviour than to change the template. It is good to follow the advice from AWS and have a hierarchy in the first place. And having that split/join would be hard to maintain.

**Comment by jufemaiz** [comment]
@yob it's more a case, imho, that @Temikus has found a bit of a "bob-each-way" position from Amazon on SSM Parameters – allowing for no slash prefix if no slashes present, but if there are, it must use an absolute Path approach with a leading slash (the hierarchy reference). Because of the bob-each-way, both `buildkite-token-default` and `/buildkite-token-default` are valid, but I can understand the approach taken by @buildkite to expect `/buildkite-token-default`.

Let's say you add some validation `AllowedPattern` for instance (ref: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#parameters-section-structure-properties) that would allow for ensuring that the Path has a leading `/` present you'd resolve the issue on the implementation expectation, and some docs on the parameter to ensure that people understand that a fully qualified path with leading slash is expected should resolve it.

**Comment by Temikus** [comment]
@jufemaiz is correct. I came from GCP to AWS recently so I didn't know that was an unspoken best practice, hence how I discovered it. We didn't have almost anything in SSM previously so I didn't have any examples to follow either. This is how I ran into this and spent quite a bit of time debugging.

I've now changed to hierarchical parameters, but leaving this here to give you the "user-journey" perspective :)
